File: typescript/src/common/input.ts
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));

function getInputPath(year: number, day: number): string {
  const dayStr = day.toString().padStart(2, '0');
  return join(__dirname, '..', '..', '..', 'inputs', `${year}`, `day${dayStr}`, 'input.txt');
}

export function input(year: number, day: number): string {
  const path = getInputPath(year, day);
  return readFileSync(path, 'utf-8').trim();
}

export function inputLines(year: number, day: number): string[] {
  return input(year, day).split('\n');
}

export function inputLinesNotBlank(year: number, day: number): string[] {
  return inputLines(year, day).filter((line) => line.trim() !== '');
}

File: rust/src/common/input.rs
use std::fs;
use std::path::PathBuf;

fn get_input_path(year: u32, day: u32) -> PathBuf {
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    PathBuf::from(manifest_dir)
        .parent()
        .unwrap()
        .join("inputs")
        .join(format!("{}", year))
        .join(format!("day{:02}", day))
        .join("input.txt")
}

pub fn input(year: u32, day: u32) -> String {
    let path = get_input_path(year, day);
    fs::read_to_string(&path)
        .unwrap_or_else(|_| panic!("Input file not found: {:?}", path))
        .trim()
        .to_string()
}

pub fn input_lines(year: u32, day: u32) -> Vec<String> {
    input(year, day).lines().map(String::from).collect()
}

pub fn input_lines_not_blank(year: u32, day: u32) -> Vec<String> {
    input_lines(year, day)
        .into_iter()
        .filter(|line| !line.trim().is_empty())
        .collect()
}
